### PR TITLE
RPC Feature Tests

### PR DESCRIPTION
**Pull request by Neurosploit**
- Added missing unit tests for the RPC feature project.
- Moved RPC integration tests to the integration test project.

### CODE REVIEW

File: src/NBitcoin/RPC/IRPCClient.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    /// <summary>
+    /// Interface for an RPC client.
+    /// </summary>
+    /// <remarks>Extend where necessary.</remarks>
+    public interface IRPCClient
+    {
+        /// <summary>Send a command over RPC.</summary>
+        /// <param name="commandName">The command to execute.</param>
+        /// <param name="parameters">An array of command parameters.</param>
+        /// <remarks>See https://en.bitcoin.it/wiki/Original_Bitcoin_client/API_calls_list </remarks>
```
Comment (noescape00) [inline src/NBitcoin/RPC/IRPCClient.cs]: Use `<see cref="https://en.bitcoin.it/wiki/Original_Bitcoin_client/API_calls_list"/>` instead.

same on other lines in this file. 

Though I think it will be better to put that link in the interface remarks instead of adding it to every single method's description.

File: src/NBitcoin/RPC/IRPCClient.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        /// <returns>The RPC response.</returns>
+        RPCResponse SendCommand(RPCOperations commandName, params object[] parameters);
+
+        /// <summary>Send a command over RPC.</summary>
+        /// <param name="rpcRequest">The rpc request to execute.</param>
```
Comment (noescape00) [inline src/NBitcoin/RPC/IRPCClient.cs]: param name does not represent the input variable name. Change here to 'request' or change 'request' to rpcRequest at 26

same on line 43

File: src/Stratis.Bitcoin.Features.RPC.Tests/Controller/RPCControllerTest.cs [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+using Stratis.Bitcoin.Features.RPC.Models;
+using Stratis.Bitcoin.Tests.Logging;
+using Stratis.Bitcoin.Utilities.JsonErrors;
+using Xunit;
+
+namespace Stratis.Bitcoin.Features.RPC.Tests.Controller
+{
+    public class RPCControllerTest : LogsTestBase, IDisposable
+    {
+        private Network network;
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.RPC.Tests/Controller/RPCControllerTest.cs]: Most of those can be made readonly or converted to a local variable.

File: src/Stratis.Bitcoin.Features.RPC/RPCClientFactory.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace Stratis.Bitcoin.Features.RPC
+{
+    /// <summary>
+    /// An interface for a factory that can create <see cref="IRPCClient"/> instances.
+    /// </summary>
+    public interface IRPCClientFactory
+    {
+        /// <summary>
+        /// Create a new RPCClient instance
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.RPC/RPCClientFactory.cs]: Missing dot here and on line 20.

File: src/Stratis.Bitcoin.Features.RPC.Tests/Controller/FullNodeControllerTest.cs [hunk 5]
```diff
@@ -0,0 +1,769 @@
...
+using Stratis.Bitcoin.Utilities;
+using Xunit;
+
+namespace Stratis.Bitcoin.Features.RPC.Tests.Controller
+{
+    public class FullNodeControllerTest : LogsTestBase, IDisposable
+    {
+        private ConcurrentChain chain;
+
+        private Mock<IFullNode> fullNode;
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.RPC.Tests/Controller/FullNodeControllerTest.cs]: Most of those can be made readonly or converted to a local variable.

File: src/Stratis.Bitcoin.Features.RPC.Tests/Controller/FullNodeControllerTest.cs [hunk 6]
```diff
@@ -0,0 +1,769 @@
...
+        }
+
+        [Fact]
+        public async Task GetRawTransactionAsync_TransactionNotInPooledTransaction_ReturnsTransactionFromBlockStoreAsync()
+        {
+            uint256 txId = new uint256(12142124);
+            this.pooledTransaction.Setup(p => p.GetTransaction(txId))
+                .ReturnsAsync((Transaction)null);
+
+            Transaction transaction = CreateTransaction();
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.RPC.Tests/Controller/FullNodeControllerTest.cs]: `this.CreateTransaction`  (also on line 142, 183, 233, 258, 345, 370, 396, 420)

File: src/Stratis.Bitcoin.Features.RPC.Tests/Controller/FullNodeControllerTest.cs [hunk 7]
```diff
@@ -0,0 +1,769 @@
...
+            return new byte[]
+            {
+                (byte)(compact >> 24),
+                (byte)(compact >> 16),
+                (byte)(compact >> 8),
+                (byte)(compact)
+            };
+        }
+
+        private class TestReadOnlyNetworkPeerCollection : IReadOnlyNetworkPeerCollection
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.RPC.Tests/Controller/FullNodeControllerTest.cs]: This class is a copy-paste from the 'WalletControllerTest'. Maybe add reference to `Stratis.Bitcoin.Features.Wallet.Tests` and reuse existing `TestReadOnlyNetworkPeerCollection`?

**Comment by Neurosploit** [comment]
Thanks for the comments @noescape00.

File: src/NBitcoin/RPC/IRPCClient.cs (hunk 1 above)
Comment (Neurosploit) [reply src/NBitcoin/RPC/IRPCClient.cs]: I will do that instead.

File: src/Stratis.Bitcoin.Features.RPC.Tests/Controller/FullNodeControllerTest.cs (hunk 7 above)
Comment (Neurosploit) [reply src/Stratis.Bitcoin.Features.RPC.Tests/Controller/FullNodeControllerTest.cs]: Deleted this instance on next commit.

File: src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs [hunk 8]
```diff
@@ -82,7 +83,10 @@ public Task Stop()
             if (!uint256.TryParse(txid, out trxid))
                 throw new ArgumentException(nameof(txid));
 
-            Transaction trx = this.pooledTransaction != null? await this.pooledTransaction.GetTransaction(trxid) : null;
+            Transaction trx = this.pooledTransaction != null ? await this.pooledTransaction.GetTransaction(trxid) : null;
+
+            if (this.pooledTransaction != null)
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs]: this code seems to be duplicated.

File: src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs [hunk 9]
```diff
@@ -214,26 +220,27 @@ public JObject ValidateAddress(string address)
             var res = new JObject();
             res["isvalid"] = false;
 
-            // P2PKH
-            if(BitcoinPubKeyAddress.IsValid(address, ref this.Network))
+            // P2WPKH
+            if (BitcoinWitPubKeyAddress.IsValid(address, ref this.Network))
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs]: why did you change the order of these?

File: src/Stratis.Bitcoin.Features.RPC/RpcSettings.cs [hunk 10]
```diff
@@ -110,26 +112,29 @@ private void CheckConfigurationValidity(ILogger logger)
 
             // We can now safely assume that server was set to true in the config or that the
             // "AddRpc" callback provided a user and password implying that the Rpc feature will be used.
-            this.Server = true;
+            if (this.RpcPassword != null && this.RpcUser != null)
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.RPC/RpcSettings.cs]: good spot.

File: src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs (hunk 9 above)
Comment (Neurosploit) [reply src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs]: I was unable to test otherwise. The tests for P2W passed on if statements before they got to P2W validations.

File: src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs (hunk 8 above)
Comment (Neurosploit) [reply src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs]: I will have a look. It seem so to me as well.

File: src/Stratis.Bitcoin.Features.RPC/RpcSettings.cs (hunk 10 above)
Comment (Neurosploit) [reply src/Stratis.Bitcoin.Features.RPC/RpcSettings.cs]: To be honest I think it's strange to set server to true automatically. This means you cannot disable the server by setting it to false in the config.

Comment (bokobza) [reply src/Stratis.Bitcoin.Features.RPC/RpcSettings.cs]: I agree, to tell the node to accept RPC command, you have to set server to true AND set the rpc fields.
I think you can remove this `this.Server = true;`.

File: src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs (hunk 9 above)
Comment (Neurosploit) [reply src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs]: Yes the ones that don't check Witness keys and scripts fail on the witness script and key with that error. I'm not sure what else I could do about it in this case. Ofcourse I could always revert the ordering change and disable the tests.

File: src/Stratis.Bitcoin.Features.RPC/RpcSettings.cs (hunk 10 above)
Comment (Neurosploit) [reply src/Stratis.Bitcoin.Features.RPC/RpcSettings.cs]: Done

**Comment by bokobza** [comment]
thanks @Neurosploit, good stuff.
